heroic-core/src/main/java/com/spotify/heroic/metric/LocalMetricManager.java

```diff
@@ -410,9 +411,43 @@ public QueryTrace buildTrace() {
                     return fullQuery;
                 })
                 .onDone(reporter.reportQueryMetrics())
-                .onDone(new EndSpanFutureReporter(findSeriesSpan));
+                .onDone(new EndSpanFutureReporter(findSeriesSpan))
+                .onDone(new FutureDone<>() {
+                    @Override
+                    public void failed(final Throwable cause) throws Exception {
+                        cleanQuotaWatchers(quotaWatcher);
+                        log.info("onDone.failed {}", quotaWatchers.size());
+                    }
+
+                    @Override
+                    public void resolved(final FullQuery result) throws Exception {
+                        cleanQuotaWatchers(quotaWatcher);
+                        log.info("onDone.resolved {}", quotaWatchers.size());
+                    }
+
+                    @Override
+                    public void cancelled() throws Exception {
+                        cleanQuotaWatchers(quotaWatcher);
+                        log.info("onDone.cancelled {}", quotaWatchers.size());
+                    }
+                });
         }
 
+        private void cleanQuotaWatchers(QuotaWatcher quotaWatcher) {
```

Comment:
This method breaks the Single Responsibility contract i.e. it performs 2 tasks - removing the supplied watcher *and* removing old watchers. Can we inline line 438? That would work.

Reply:
Done